update/release.go

```diff
@@ -59,6 +73,240 @@ func (s ReleaseSpec) ReleaseType() string {
+	// Compile an `ImageMap` of all relevant images
+	var images ImageMap
+	var repo string
+	var err error
+
+	switch s.ImageSpec {
+	case ImageSpecLatest:
+		images, err = collectUpdateImages(rc.Registry(), candidates)
+	default:
+		var image flux.ImageID
+		image, err = s.ImageSpec.AsID()
+		if err == nil {
+			repo = image.Repository()
+			images, err = exactImages(rc.Registry(), []flux.ImageID{image})
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Look through all the services' containers to see which have an
+	// image that could be updated.
+	var updates []*ServiceUpdate
+	for _, u := range candidates {
+		containers, err := u.Service.ContainersOrError()
+		if err != nil {
+			results[u.ServiceID] = ServiceResult{
+				Status: ReleaseStatusFailed,
+				Error:  err.Error(),
+			}
+			continue
+		}
+
+		// If at least one container used an image in question, we say
+		// we're skipping it rather than ignoring it. This is mainly
+		// for the purpose of filtering the output.
+		ignoredOrSkipped := ReleaseStatusIgnored
+		var containerUpdates []ContainerUpdate
+
+		for _, container := range containers {
+			currentImageID, err := flux.ParseImageID(container.Image)
+			if err != nil {
+				// We may hope never to find a malformed image ID, but
+				// anything is possible.
+				return nil, err
+			}
+
+			latestImage := images.LatestImage(currentImageID.Repository(), "*")
+			if latestImage == nil {
+				if currentImageID.Repository() != repo {
+					ignoredOrSkipped = ReleaseStatusIgnored
+				} else {
+					ignoredOrSkipped = ReleaseStatusUnknown
+				}
+				continue
+			}
+
+			if currentImageID == latestImage.ID {
+				ignoredOrSkipped = ReleaseStatusSkipped
+				continue
+			}
+
+			u.ManifestBytes, err = rc.Manifests().UpdateDefinition(u.ManifestBytes, container.Name, latestImage.ID)
+			if err != nil {
+				return nil, err
+			}
+
+			containerUpdates = append(containerUpdates, ContainerUpdate{
+				Container: container.Name,
+				Current:   currentImageID,
+				Target:    latestImage.ID,
+			})
+		}
+
+		switch {
+		case len(containerUpdates) > 0:
+			u.Updates = containerUpdates
+			updates = append(updates, u)
+			results[u.ServiceID] = ServiceResult{
+				Status:       ReleaseStatusSuccess,
+				PerContainer: containerUpdates,
+			}
+		case ignoredOrSkipped == ReleaseStatusSkipped:
+			results[u.ServiceID] = ServiceResult{
+				Status: ReleaseStatusSkipped,
+				Error:  ImageUpToDate,
+			}
+		case ignoredOrSkipped == ReleaseStatusIgnored:
+			results[u.ServiceID] = ServiceResult{
+				Status: ReleaseStatusIgnored,
+				Error:  DoesNotUseImage,
+			}
+		case ignoredOrSkipped == ReleaseStatusUnknown:
+			results[u.ServiceID] = ServiceResult{
+				Status: ReleaseStatusSkipped,
+				Error:  ImageNotFound,
+			}
+		}
+	}
+
+	return updates, nil
+}
+
+func (s ReleaseSpec) Observe(start time.Time, err error) {
```

Comment:
I didn't realise that was an empty method. Must be more careful with my commits.